### PR TITLE
Xyuan/p3 ice collection

### PR DESCRIPTION
**Pull request by xyuan**
Redo this p3_ice_collection function porting. It includes porting
ice_cldliq_collection
ice_rain_collection
ice_self_collection
to c++.

BFB tests in summit passed.

[BFB]

### CODE REVIEW

**Comment by xyuan** [comment]
This PR is ready for review, please review it.

File: components/scream/src/physics/p3/p3_functions_ice_collection_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+   constexpr auto inv_dropmass = C::ONE/C::dropmass;
+ 
+   qccol.set(qitot_incld_gt_small && qc_incld_gt_small && t_is_negative, 
+             rhofaci*f1pr04*qc_incld*eci*rho*nitot_incld);
+   nccol.set(qitot_incld_gt_small && qc_incld_gt_small, 
+             rhofaci*f1pr04*nc_incld*eci*rho*nitot_incld);
+
+   // for T > 273.15, assume cloud water is collected and shed as rain drops
+   // sink for cloud water mass and number, note qcshed is source for rain mass        
+   qcshd.set(qitot_incld_gt_small && qc_incld_gt_small && !t_is_negative,
```
Comment (ambrad) [inline components/scream/src/physics/p3/p3_functions_ice_collection_impl.hpp]: Any mask that is reused should be saved in a const auto temporary.

**Comment by jgfouca** [comment]
@ambrad , how's that look?

**Comment by jgfouca** [comment]
Just making a note that the fortran changes in this PR are not BFB with master fortran due to using the CXX bfb math layer.

**Pull request merged**